### PR TITLE
Add experimental batch support

### PR DESCRIPTION
**Pull request by mrusu91**

### CODE REVIEW

**Comment by JaimeLennox** [comment]
@aschmolck I'm reasonably happy with the state of this now.

There's quite a major refactor involved here, in order to keep as much functionality as possible between the regular job and the batch job. As such, the tests have been shuffled around quite a lot too.

I'm entirely happy with the state of the batch job tests, as there are some core tests missing here, but to be honest I do find the mock API setup a bit confusing and I think it'd be nicer to get this merged and improve on that later, especially as this functionality is considered experimental for now.

**Comment by JaimeLennox** [comment]
Rebased. What are the typos you want me to fix?

Regarding the history, I actually think the commits are separate enough to warrant the number, unless you want a big super-commit - but seeing as the big refactor is halfway through, I'm not sure this makes sense. If you have any suggestions for commits to squash I'd appreciate it.

I agree that using the state machine would've been the nicest approach, but I had a hard time getting my head around it. This was the easier method, for now, and we can probably work towards improving the tests later.

**Comment by aschmolck** [comment]
Super-commit: I fully agree that it's not desirable to squash the mid-way refactoring nor the refinements such as the job cancellation or the modification to the batching logic away, just intermediate fixups to these (fix tests; rebase fixes etc). I have annotated the individual commits that I think are candidates as such.

Typos: `---batch`, noted above; I thought there was another in a commit message, but since I can't see it anymore, maybe I misremembered.

Tests: agreed.

File: marge/batch_job.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+            'labels': BatchMergeJob.BATCH_BRANCH_NAME,
+            'state': 'opened',
+            'order_by': 'created_at',
+            'sort': 'desc',
+        }
+        batch_mrs = MergeRequest.search(
+            api=self._api,
+            project_id=self._project.id,
+            params=params,
+        ) or []
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: Correct. Will fix.

File: marge/batch_job.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+    def ensure_mr_not_changed(self, merge_request):
+        log.info('Ensuring MR !%s did not change', merge_request.iid)
+        changed_mr = MergeRequest.fetch_by_iid(
+            merge_request.project_id,
+            merge_request.iid,
+            self._api,
+        )
+        error_message = 'The {} changed whilst merging!'
+        if changed_mr.source_branch != merge_request.source_branch:
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: marge/git.py [hunk 3]
```diff
@@ -109,10 +124,12 @@ def _fuse_branch(self, strategy, branch, target_branch, source_repo_url=None):
...
     def remove_branch(self, branch, *, new_current_branch='master'):
         assert branch != new_current_branch
-        self.git('checkout', new_current_branch, '--')
         self.git('branch', '-D', branch)
 
-    def push_force(self, branch, source_repo_url=None):
+    def checkout_branch(self, branch, start_point=''):
+        self.git('checkout', '-B', branch, start_point, '--')
+
+    def push(self, branch, source_repo_url=None, force=False):
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/git.py]: :+1:

File: marge/git.py [hunk 4]
```diff
@@ -126,7 +143,11 @@ def push_force(self, branch, source_repo_url=None):
             source = 'source'
         else:
             source = 'origin'
-        self.git('push', '--force', source, branch)
+        force_flag = '--force' if force else ''
+        self.git('push', force_flag, source, '%s:%s' % (branch, branch))
+
+    def push_force(self, branch, source_repo_url=None):
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/git.py]: Yep, not sure why it really existed in the first place.

File: marge/batch_job.py [hunk 5]
```diff
@@ -15,7 +15,7 @@
     'WIP': 'Is marked as Work-In-Progress.',
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: marge/merge_request.py [hunk 6]
```diff
@@ -126,9 +126,10 @@ def accept(self, remove_branch=False, sha=None):
             ),
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/merge_request.py]: Yeah, I'd rather keep this separate to distinguish my commits from those of @mrusu91

File: tests/test_batch_job.py [hunk 7]
```diff
@@ -223,67 +230,37 @@ def test_ensure_mergeable_mr_ci_not_ok(bmj_get_mr_ci_status):
...
     merge_request.fetch_approvals.return_value.sufficient = True
-    with pytest.raises(MergeError) as exc_info:
+    with pytest.raises(CannotMerge) as exc_info:
         batch_merge_job.ensure_mergeable_mr(merge_request)
 
-    assert exc_info.value.message_key == 'CI_NOT_OK'
+    assert exc_info.value.reason == 'This MR has not passed CI'
 
 
-@patch.object(BatchMergeJob, 'get_source_project')
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply tests/test_batch_job.py]: I think these are now part of `test_job.py`, which has the generic functionality anyway.

File: marge/batch_job.py [hunk 8]
```diff
@@ -179,41 +179,48 @@ def execute(self):
         self._repo.checkout_branch(target_branch, 'origin/%s' % target_branch)
         self._repo.checkout_branch(BatchMergeJob.BATCH_BRANCH_NAME, 'origin/%s' % target_branch)
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: tests/test_job.py [hunk 9]
```diff
@@ -395,7 +395,7 @@ def test_tells_explicitly_that_gitlab_refused_to_merge(self, unused_time_sleep):
                 dict(sha=rewritten_sha, should_remove_source_branch=True, merge_when_pipeline_succeeds=True),
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply tests/test_job.py]: :+1:

File: marge/batch_job.py [hunk 10]
```diff
@@ -86,7 +86,6 @@ def get_mergeable_mrs(self, merge_requests):
                 mergeable_mrs.append(merge_request)
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: tests/test_job.py [hunk 11]
```diff
@@ -181,15 +182,19 @@ def make_job(self, options=None):
...
+        return marge.single_merge_job.SingleMergeJob(
             api=api, user=user,
             project=project, merge_request=merge_request, repo=repo,
             options=options,
         )
 
     def test_succeeds_first_time(self, unused_time_sleep):
         api, mocklab = self.api, self.mocklab
-        with patch('marge.job.update_from_target_branch_and_push', side_effect=mocklab.push_updated):
+        with patch.object(
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply tests/test_job.py]: :+1:

File: README.md [hunk 12]
```diff
@@ -105,6 +105,8 @@ optional arguments:
                            [env var: MARGE_BRANCH_REGEXP] (default: .*)
   --debug               Debug logging (includes all HTTP requests etc).
                            [env var: MARGE_DEBUG] (default: False)
+  ---batch  Enable processing MRs in batches.
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply README.md]: :+1:

File: marge/batch_job.py [hunk 13]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=too-many-branches,too-many-statements
 import logging as log
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: marge/batch_job.py [hunk 14]
```diff
@@ -229,6 +229,8 @@ def execute(self):
                 continue
             else:
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: marge/batch_job.py [hunk 15]
```diff
@@ -219,6 +219,8 @@ def execute(self):
             batch_mr = self.create_batch_mr(
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: I thought it better to leave a rename commit as separate, as not to mess with the history too much (and to make it easier to `git blame` in the future.

File: marge/batch_job.py [hunk 16]
```diff
@@ -220,8 +220,18 @@ def execute(self):
                 target_branch=target_branch,
             )
             for merge_request in working_merge_requests:
-                merge_request.comment('Attempting to batch as part of !{}'.format(batch_mr.iid))
-            self.wait_for_ci_to_pass(batch_mr)
+                merge_request.comment('I will attempt to batch this MR (!{})...'.format(batch_mr.iid))
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/batch_job.py]: :+1:

File: marge/pipeline.py [hunk 17]
```diff
@@ -1,19 +1,36 @@
...
 
 
 class Pipeline(gitlab.Resource):
+    def __init__(self, api, info, project_id):
+        info['project_id'] = project_id
+        super().__init__(api, info)
 
     @classmethod
-    def pipelines_by_branch(cls, project_id, branch, api):
+    def pipelines_by_branch(cls, project_id, branch, api, **extra_params):
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/pipeline.py]: Ok, I've changed.

File: marge/job.py [hunk 18]
```diff
@@ -5,7 +5,6 @@
 from datetime import datetime, timedelta
```
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/job.py]: I appear to have lost which commit this refers to.

**Comment by JaimeLennox** [comment]
@aschmolck I've made the changes requested, including squashing the intermediate commits. Could you have another look and check everything is ok now?

File: marge/job.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (JaimeLennox) [reply marge/job.py]: Actually, if this was a reference to the "rebase fixes" commit, then I've since squashed that.

**Comment by aschmolck** [comment]
@JaimeLennox great -- merged! Thanks, and sorry for the wait!

**Pull request merged**